packages/plugin-backgroundcolor/src/BackgroundColor.ts

```diff
@@ -0,0 +1,164 @@
+        if (backgroundColor instanceof VNode) {
+            node = backgroundColor;
+        }
+        const nodeBackgroundColor = getStyle(node, 'background-color');
+        if (!nodeBackgroundColor || backgroundColor instanceof VNode) {
+            return !!nodeBackgroundColor;
+        } else {
+            return nodeBackgroundColor === backgroundColor;
+        }
+    }
+    static dependencies = [Inline];
+    configuration = { ...defaultConfiguration, ...this.configuration };
+    commands = {
+        colorBackground: {
+            handler: this.colorBackground,
+        },
+        uncolorBackground: {
+            handler: this.uncolorBackground,
+        },
+    };
+    readonly loadables: Loadables<Keymap> = {
+        shortcuts: [
+            {
+                pattern: 'CTRL+H',
+                commandId: 'colorBackground',
+                // TODO: use dialog to get params
+                commandArgs: {
+                    color: 'yellow',
+                },
+            },
+            {
+                pattern: 'CTRL+SHIFT+H',
+                commandId: 'uncolorBackground',
+            },
+        ],
+    };
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * Apply the given background color to the range.
+     *
+     * @param params
+     */
+    colorBackground(params: ColorParams): void {
+        const color = params.color;
+
+        if (params.context.range.isCollapsed()) {
+            // Set the style cache.
+            const inline = this.dependencies.get(Inline);
+            const currentCache = inline.styleCache || {};
+            inline.styleCache = { ...currentCache, 'background-color': color };
+        } else {
+            for (const node of this._highestSelectedNodes(params.context.range)) {
+                // Apply the style to the node or its first format.
+                setStyle(this._nodeOrFirstFormat(node), 'background-color', color);
+
+                // If there are ancestors of this node whose children all have
+                // this style, style these ancestors instead of their
+                // descendants.
+                let parent = node.parent;
+                while (parent?.children().every(child => BackgroundColor.hasColor(color, child))) {
+                    setStyle(parent, 'background-color', color);
+                    for (const child of parent.children()) {
+                        removeStyle(child, 'background-color');
+                    }
+                    parent = parent.parent;
+                }
+            }
+        }
+    }
+    /**
+     * Remove the current background color from the range. If the color was
+     * applied to an ancestor, apply the default color to its relevant inline
+     * descendants.
+     *
+     * @param params
+     */
+    uncolorBackground(params: CommandParams): void {
```

Comment:
Whether it is `uncolorBackground` or `resetBackgroundColor` or whatever, I think it should not set `style=background-color: defaultColor` if it did not originally had a color. I fear that, currently, if i have:
```html
<p>[abcde]</p>
``` 
I do color: red, I get:
```html
<p style="backround-color:red">abcde</p>
```
Alright. Now I select only some chars:
```html
<p  style="backround-color:red">ab[cd]e</p>
```
I do uncolor, I think I'll get:
```html
<p  style="backround-color:red">ab<span style="backround-color:white">[cd]</span>e</p>
```
Isn't that weird ? `background-color: white` ? As a user, I expected `resetBackgroundColor` or `uncolorBackground` to somewhat *undo* the coloration I did on these characters, not *adding* yet another coloring that *happens* to be the one I had before. I'm not sure what to expect in this case though, because I guess I expected this:
```html
<p><span style="backround-color:red">ab</span>[cd]<span style="backround-color:red">e</span></p>
```
Which would also be weird I suppose, since it kinda "undid" the style that was on the `p`. Remind me a bit of what was the point of handling the style on the container ? It ends up making both scenarious looking weird at the end.. :/

Reply:
Both scenarios _are_ weird I'm afraid... Whatever we do we could parse parents with the style so we have to handle it.

Reply:
My point was that it wouldn't have been weird if the first `color` command had not been applied on the container but on the selected children instead:
```html
<p><span style="backround-color:red">abcde</span></p>
```
Then the second `color` would result in:
```html
<p><span style="backround-color:red">ab</span>[cd]<span style="backround-color:red">e</span></p>
```
Which would not be weird this time since the style was never applied to the parent to begin with. Suboptimal, maybe, but not weird.

I guess you're right with the style on the parent from the original parsing though.